lab_4/main.py

```diff
@@ -1,74 +1,232 @@
+                context_freq += freq
+
+        if context_freq:
+            context_freq = self._n_gram_trie.n_gram_frequencies.get(context + (word,), 0) / context_freq
+
+        return context_freq
 
     def _generate_next_word(self, context: tuple) -> int:
+
+        if (not isinstance(context, tuple) or
+                not context or
+                len(context) != self._n_gram_trie.size - 1):
+            raise ValueError
+
+        max_freq, word = 0, 0
+
+        for value in self._word_storage.storage.values():
+            frequency = self._calculate_maximum_likelihood(value, context)
+
+            if frequency > max_freq:
+                max_freq, word = frequency, value
+
+        if not max_freq:
+            word = max(self._n_gram_trie.uni_grams, key=self._n_gram_trie.uni_grams.get)[0]
+
+        return word
+
+    def sample_method(self):
         pass
 
 
 class BackOffGenerator(NGramTextGenerator):
 
     def __init__(self, word_storage: WordStorage, n_gram_trie: NGramTrie, *args):
         super().__init__(word_storage, n_gram_trie)
+        self._n_gram_tries = (n_gram_trie,) + args
 
     def _generate_next_word(self, context: tuple) -> int:
-        pass
+        if (not isinstance(context, tuple) or
+                len(context) + 1 != self._n_gram_trie.size or
+                context[0] > len(self._word_storage.storage)):
+            raise ValueError
+
+        max_size = len(context) + 1
+        self._n_gram_tries = [trie for trie in self._n_gram_tries if trie.size <= max_size]
+
+        for trie in self._n_gram_tries:
+            self._n_gram_trie = trie
+
+            if n_gram := super()._generate_next_word(context):
+                return n_gram
+            context = context[:-1]
+
+        return n_gram
 
 
 def decode_text(storage: WordStorage, encoded_text: tuple) -> tuple:
-    pass
+    if not isinstance(storage, WordStorage) or not isinstance(encoded_text, tuple):
+        raise ValueError
+
+    text = ' '.join(storage.get_word(word) for word in encoded_text) + ' '
+    text = text.split(' <END> ')
+    text = tuple(sent.capitalize() for sent in text)
+
+    return text[:-1]
 
 
 def save_model(model: NGramTextGenerator, path_to_saved_model: str):
-    pass
+    if not isinstance(model, NGramTextGenerator) or not isinstance(path_to_saved_model, str):
+        raise ValueError
+
+    with open(path_to_saved_model, 'w', encoding='utf-8') as obj:
+        model_attrs = {x[0]: x[1].__dict__ for x in model.__dict__.items()}
+        obj.write(str(model_attrs))
 
 
 def load_model(path_to_saved_model: str) -> NGramTextGenerator:
-    pass
+    if not isinstance(path_to_saved_model, str):
+        raise ValueError
+
+    with open(path_to_saved_model) as obj:
+        attrs = dict(ast.literal_eval(obj.read()))
+
+    storage = type('WordStorage', (), attrs['_word_storage'])
+    trie = type('NGramTrie', (), attrs['_n_gram_trie'])
+
+    class Model(NGramTextGenerator):
```

Comment:
why do we need to create class here?

Reply:
Well, we actually don't, I just forgot to remove these lines after changing the serializiation strategy)